### ISSUE TITLE
Invalid version string on Update

### ISSUE THREAD

**Issue by cpaulik** [syl20bnr/spacemacs#839]
I get the following error when pressing the Update button. This might be a MELPA issue?

```
Debugger entered--Lisp error: (error "Invalid version string: 'nil'")
  signal(error ("Invalid version string: 'nil'"))
  error("Invalid version string: '%s'" nil)
  version-to-list(nil)
  version<=(nil "20120402")
  (or (null installed-version) (version<= newest-version installed-version))
  (if (or (null installed-version) (version<= newest-version installed-version)) nil (if (member pkg result) result (setq result (append result (list pkg)))))
  (let ((installed-version (configuration-layer//get-package-version-string pkg)) (newest-version (configuration-layer//get-latest-package-version-string pkg))) (if (or (null installed-version) (version<= newest-version installed-version)) nil (if (member pkg result) result (setq result (append result (list pkg))))))
  (while --dolist-tail-- (setq pkg (car --dolist-tail--)) (let* ((deps (configuration-layer//get-package-dependencies-from-archive pkg)) (update-deps (if deps (progn (configuration-layer//get-packages-to-update (mapcar ... deps)))))) (if update-deps (progn (setq result (append update-deps result))))) (let ((installed-version (configuration-layer//get-package-version-string pkg)) (newest-version (configuration-layer//get-latest-package-version-string pkg))) (if (or (null installed-version) (version<= newest-version installed-version)) nil (if (member pkg result) result (setq result (append result (list pkg)))))) (setq --dolist-tail-- (cdr --dolist-tail--)))
  (let ((--dolist-tail-- packages) pkg) (while --dolist-tail-- (setq pkg (car --dolist-tail--)) (let* ((deps (configuration-layer//get-package-dependencies-from-archive pkg)) (update-deps (if deps (progn (configuration-layer//get-packages-to-update ...))))) (if update-deps (progn (setq result (append update-deps result))))) (let ((installed-version (configuration-layer//get-package-version-string pkg)) (newest-version (configuration-layer//get-latest-package-version-string pkg))) (if (or (null installed-version) (version<= newest-version installed-version)) nil (if (member pkg result) result (setq result (append result (list pkg)))))) (setq --dolist-tail-- (cdr --dolist-tail--))))
  (let (result) (let ((--dolist-tail-- packages) pkg) (while --dolist-tail-- (setq pkg (car --dolist-tail--)) (let* ((deps (configuration-layer//get-package-dependencies-from-archive pkg)) (update-deps (if deps (progn ...)))) (if update-deps (progn (setq result (append update-deps result))))) (let ((installed-version (configuration-layer//get-package-version-string pkg)) (newest-version (configuration-layer//get-latest-package-version-string pkg))) (if (or (null installed-version) (version<= newest-version installed-version)) nil (if (member pkg result) result (setq result (append result ...))))) (setq --dolist-tail-- (cdr --dolist-tail--)))) (delete-dups result))
  (progn (let (result) (let ((--dolist-tail-- packages) pkg) (while --dolist-tail-- (setq pkg (car --dolist-tail--)) (let* ((deps (configuration-layer//get-package-dependencies-from-archive pkg)) (update-deps (if deps ...))) (if update-deps (progn (setq result ...)))) (let ((installed-version (configuration-layer//get-package-version-string pkg)) (newest-version (configuration-layer//get-latest-package-version-string pkg))) (if (or (null installed-version) (version<= newest-version installed-version)) nil (if (member pkg result) result (setq result ...)))) (setq --dolist-tail-- (cdr --dolist-tail--)))) (delete-dups result)))
  (if packages (progn (let (result) (let ((--dolist-tail-- packages) pkg) (while --dolist-tail-- (setq pkg (car --dolist-tail--)) (let* ((deps ...) (update-deps ...)) (if update-deps (progn ...))) (let ((installed-version ...) (newest-version ...)) (if (or ... ...) nil (if ... result ...))) (setq --dolist-tail-- (cdr --dolist-tail--)))) (delete-dups result))))
  configuration-layer//get-packages-to-update((ac-anaconda ace-jump-mode ace-window afternoon-theme ag aggressive-indent alect-themes ample-theme ample-zen-theme anaconda-mode anti-zenburn-theme async auctex auto-dictionary auto-highlight-symbol base16-theme birds-of-paradise-plus-theme bookmark bubbleberry-theme buffer-move busybee-theme cherry-blossom-theme clues-theme color-theme-sanityinc-solarized color-theme-sanityinc-tomorrow color-theme-solarized colorsarenice-theme company company-anaconda company-auctex company-c-headers company-quickhelp company-tern cyberpunk-theme cython-mode dakrone-theme darkburn-theme darkmine-theme dash diminish dired+ django-theme doc-view ediff eldoc elisp-slime-nav emmet-mode espresso-theme evil evil-anzu ...))
  (let* ((update-packages (configuration-layer//get-packages-to-update configuration-layer-all-packages-sorted)) (date (format-time-string "%y-%m-%d_%H.%M.%S")) (rollback-dir (expand-file-name (concat configuration-layer-rollback-directory (file-name-as-directory date)))) (upgrade-count (length update-packages)) (upgraded-count 0) (update-packages-alist)) (if (> upgrade-count 0) (if (not (yes-or-no-p (format (concat "%s package(s) to update, " "do you want to continue ? ") upgrade-count))) (spacemacs/append-to-buffer "Packages update has been cancelled.\n") (spacemacs/replace-last-line-of-buffer "--> performing backup of package(s) to update...\n" t) (spacemacs//redisplay) (let ((--dolist-tail-- update-packages) pkg) (while --dolist-tail-- (setq pkg (car --dolist-tail--)) (let* ((src-dir ...) (dest-dir ...)) (copy-directory src-dir dest-dir (quote keeptime) (quote create) (quote copy-content)) (setq update-packages-alist (cons ... update-packages-alist))) (setq --dolist-tail-- (cdr --dolist-tail--)))) (spacemacs/dump-vars-to-file (quote (update-packages-alist)) (expand-file-name (concat rollback-dir configuration-layer-rollback-info))) (let ((--dolist-tail-- update-packages) pkg) (while --dolist-tail-- (setq pkg (car --dolist-tail--)) (setq upgraded-count (1+ upgraded-count)) (spacemacs/replace-last-line-of-buffer (format "--> updating package %s... [%s/%s]" pkg upgraded-count upgrade-count) t) (spacemacs//redisplay) (configuration-layer//package-delete pkg) (condition-case err (package-install pkg) ((quote error) (message ...) (package-install pkg))) (if (version< emacs-version "24.3.50") (progn (setq package-activated-list ...) (configuration-layer//activate-package pkg))) (setq --dolist-tail-- (cdr --dolist-tail--)))) (spacemacs/append-to-buffer (format "\n--> %s packages updated.\n" upgraded-count)) (spacemacs/append-to-buffer "\nEmacs has to be restarted for the changes to take effect.\n") (spacemacs//redisplay)) (spacemacs/append-to-buffer "--> All packages are up to date.\n") (spacemacs//redisplay)))
  configuration-layer/update-packages()
  (lambda (b) (configuration-layer/update-packages))(#<overlay from 200 to 208 in *spacemacs*>)
  button-activate(#<overlay from 200 to 208 in *spacemacs*> nil)
  push-button(205)
  push-button((mouse-2 (#<window 3 on *spacemacs*> 205 (687 . 381) 672524 nil 205 (52 . 5) nil (11 . 13) (0 . 28))))
  call-interactively(push-button nil nil)
  command-execute(push-button)

```

**Comment by syl20bnr** [comment]
I hope to be able to reproduce it. This is like a built-in package (version is nil for them) has been moved to MELPA :-)

**Comment by codygman** [comment]
I'm also getting the package version is nil error, I'm on v0.101.1.

**Comment by sooheon** [comment]
Same issue here on v0.101.2, I get `version-to-list: Invalid version string: 'nil'` in my messages. @codygman, you can manually update for now by SPC-a-P, U, x (using Paradox).

**Comment by emmanueltouzery** [comment]
FYI just happened to me. The trick from @sooheon appears to work.

**Comment by autosquid** [comment]
I met the same problem. Is there any way to find out which damn package is it ?

**Comment by kccai** [comment]
I am on today's develop branch. I have this problem too and find the "version-to-list: Invalid version string: 'nil' " in the Message buffer.

**Comment by gavocanov** [comment]
same here

**Comment by bmag** [comment]
I'm encountering the same issue now. The backtrace has changed since @cpaulik has posted in March:
```
Debugger entered--Lisp error: (error "Invalid version string: 'nil'")
  signal(error ("Invalid version string: 'nil'"))
  error("Invalid version string: '%s'" nil)
  version-to-list(nil)
  version<("20150307.208" nil)
  (progn (setq new-version (if recipe (quelpa-checkout recipe (expand-file-name (symbol-name pkg-name) quelpa-build-dir)) (configuration-layer//get-latest-package-version-string pkg-name))) (version< cur-version new-version))
  (if cur-version (progn (setq new-version (if recipe (quelpa-checkout recipe (expand-file-name (symbol-name pkg-name) quelpa-build-dir)) (configuration-layer//get-latest-package-version-string pkg-name))) (version< cur-version new-version)))
  (let ((recipe (configuration-layer//get-package-recipe pkg-name)) (cur-version (configuration-layer//get-package-version-string pkg-name)) new-version) (if cur-version (progn (setq new-version (if recipe (quelpa-checkout recipe (expand-file-name (symbol-name pkg-name) quelpa-build-dir)) (configuration-layer//get-latest-package-version-string pkg-name))) (version< cur-version new-version))))
  configuration-layer//new-version-available-p(highlight-indentation)
  funcall(configuration-layer//new-version-available-p highlight-indentation)
  (if (funcall filter pkg-name) (progn (if (member pkg-name result) result (setq result (append result (list pkg-name))))))
  (while --dolist-tail-- (setq pkg-name (car --dolist-tail--)) (let* ((deps (configuration-layer//get-package-deps-from-archive pkg-name)) (install-deps (if deps (progn (configuration-layer//filter-packages-with-deps (mapcar ... deps) filter))))) (if install-deps (progn (setq result (append install-deps result))))) (if (funcall filter pkg-name) (progn (if (member pkg-name result) result (setq result (append result (list pkg-name)))))) (setq --dolist-tail-- (cdr --dolist-tail--)))
  (let ((--dolist-tail-- pkg-names) pkg-name) (while --dolist-tail-- (setq pkg-name (car --dolist-tail--)) (let* ((deps (configuration-layer//get-package-deps-from-archive pkg-name)) (install-deps (if deps (progn (configuration-layer//filter-packages-with-deps ... filter))))) (if install-deps (progn (setq result (append install-deps result))))) (if (funcall filter pkg-name) (progn (if (member pkg-name result) result (setq result (append result (list pkg-name)))))) (setq --dolist-tail-- (cdr --dolist-tail--))))
  (let (result) (let ((--dolist-tail-- pkg-names) pkg-name) (while --dolist-tail-- (setq pkg-name (car --dolist-tail--)) (let* ((deps (configuration-layer//get-package-deps-from-archive pkg-name)) (install-deps (if deps (progn ...)))) (if install-deps (progn (setq result (append install-deps result))))) (if (funcall filter pkg-name) (progn (if (member pkg-name result) result (setq result (append result ...))))) (setq --dolist-tail-- (cdr --dolist-tail--)))) (delete-dups result))
  (progn (let (result) (let ((--dolist-tail-- pkg-names) pkg-name) (while --dolist-tail-- (setq pkg-name (car --dolist-tail--)) (let* ((deps (configuration-layer//get-package-deps-from-archive pkg-name)) (install-deps (if deps ...))) (if install-deps (progn (setq result ...)))) (if (funcall filter pkg-name) (progn (if (member pkg-name result) result (setq result ...)))) (setq --dolist-tail-- (cdr --dolist-tail--)))) (delete-dups result)))
  (if pkg-names (progn (let (result) (let ((--dolist-tail-- pkg-names) pkg-name) (while --dolist-tail-- (setq pkg-name (car --dolist-tail--)) (let* ((deps ...) (install-deps ...)) (if install-deps (progn ...))) (if (funcall filter pkg-name) (progn (if ... result ...))) (setq --dolist-tail-- (cdr --dolist-tail--)))) (delete-dups result))))
  configuration-layer//filter-packages-with-deps((ac-ispell ace-link ace-window adaptive-wrap afternoon-theme aggressive-indent alect-themes align-cljlet ample-theme ample-zen-theme anaconda-mode anti-zenburn-theme apropospriate-theme auto-complete auto-dictionary auto-highlight-symbol auto-yasnippet avy bind-key birds-of-paradise-plus-theme bookmark bubbleberry-theme buffer-move busybee-theme cc-mode cherry-blossom-theme cider cider-eval-sexp-fu clang-format clean-aindent-mode clj-refactor clojure-mode clues-theme cmake-mode coffee-mode color-theme-sanityinc-solarized color-theme-sanityinc-tomorrow colorsarenice-theme command-log-mode company company-anaconda company-c-headers company-quickhelp company-statistics company-tern company-web css-mode cyberpunk-theme cython-mode dakrone-theme ...) configuration-layer//new-version-available-p)
  configuration-layer//get-packages-to-update((ac-ispell ace-link ace-window adaptive-wrap afternoon-theme aggressive-indent alect-themes align-cljlet ample-theme ample-zen-theme anaconda-mode anti-zenburn-theme apropospriate-theme auto-complete auto-dictionary auto-highlight-symbol auto-yasnippet avy bind-key birds-of-paradise-plus-theme bookmark bubbleberry-theme buffer-move busybee-theme cc-mode cherry-blossom-theme cider cider-eval-sexp-fu clang-format clean-aindent-mode clj-refactor clojure-mode clues-theme cmake-mode coffee-mode color-theme-sanityinc-solarized color-theme-sanityinc-tomorrow colorsarenice-theme command-log-mode company company-anaconda company-c-headers company-quickhelp company-statistics company-tern company-web css-mode cyberpunk-theme cython-mode dakrone-theme ...))
  (let* ((candidates (configuration-layer/filter-packages configuration-layer-packages (function (lambda (x) (and (not ...) (not ...) (not ...)))))) (update-packages (configuration-layer//get-packages-to-update (mapcar (quote car) (object-assoc-list :name candidates)))) (date (format-time-string "%y-%m-%d_%H.%M.%S")) (rollback-dir (expand-file-name (concat configuration-layer-rollback-directory (file-name-as-directory date)))) (upgrade-count (length update-packages)) (upgraded-count 0) (update-packages-alist)) (if (> upgrade-count 0) (if (and (not always-update) (not (yes-or-no-p (format (concat "%s package(s) to update, " "do you want to continue ? ") upgrade-count)))) (spacemacs-buffer/append "Packages update has been cancelled.\n") (spacemacs-buffer/append "--> performing backup of package(s) to update...\n" t) (spacemacs//redisplay) (let ((--dolist-tail-- update-packages) pkg) (while --dolist-tail-- (setq pkg (car --dolist-tail--)) (let* ((src-dir ...) (dest-dir ...)) (copy-directory src-dir dest-dir (quote keeptime) (quote create) (quote copy-content)) (setq update-packages-alist (cons ... update-packages-alist))) (setq --dolist-tail-- (cdr --dolist-tail--)))) (spacemacs/dump-vars-to-file (quote (update-packages-alist)) (expand-file-name (concat rollback-dir configuration-layer-rollback-info))) (let ((--dolist-tail-- update-packages) pkg) (while --dolist-tail-- (setq pkg (car --dolist-tail--)) (setq upgraded-count (1+ upgraded-count)) (spacemacs-buffer/replace-last-line (format "--> preparing update of package %s... [%s/%s]" pkg upgraded-count upgrade-count) t) (spacemacs//redisplay) (configuration-layer//package-delete pkg) (setq --dolist-tail-- (cdr --dolist-tail--)))) (spacemacs-buffer/append (format "\n--> %s package(s) to be updated.\n" upgraded-count)) (spacemacs-buffer/append "\nEmacs has to be restarted to actually install the new packages.\n") (spacemacs//redisplay)) (spacemacs-buffer/append "--> All packages are up to date.\n") (spacemacs//redisplay)))
  configuration-layer/update-packages()
  (lambda (&rest ignore) (configuration-layer/update-packages))((push-button :args nil :value #("Update" 0 6 (face font-lock-keyword-face)) :help-echo "Update all ELPA packages to the latest versions." :action (lambda (&rest ignore) (configuration-layer/update-packages)) :mouse-face highlight :follow-link "
" :button-overlay #<overlay from 197 to 205 in *spacemacs*> :from #<marker (moves after insertion) at 197 in *spacemacs*> :to #<marker at 205 in *spacemacs*>) nil)
  widget-apply((push-button :args nil :value #("Update" 0 6 (face font-lock-keyword-face)) :help-echo "Update all ELPA packages to the latest versions." :action (lambda (&rest ignore) (configuration-layer/update-packages)) :mouse-face highlight :follow-link "
" :button-overlay #<overlay from 197 to 205 in *spacemacs*> :from #<marker (moves after insertion) at 197 in *spacemacs*> :to #<marker at 205 in *spacemacs*>) :action nil)
  widget-apply-action((push-button :args nil :value #("Update" 0 6 (face font-lock-keyword-face)) :help-echo "Update all ELPA packages to the latest versions." :action (lambda (&rest ignore) (configuration-layer/update-packages)) :mouse-face highlight :follow-link "
" :button-overlay #<overlay from 197 to 205 in *spacemacs*> :from #<marker (moves after insertion) at 197 in *spacemacs*> :to #<marker at 205 in *spacemacs*>) nil)
  widget-button-press(198)
  (progn (goto-char (1+ res)) (widget-button-press (point)))
  (if res (progn (goto-char (1+ res)) (widget-button-press (point))))
  (let ((res (avy--with-avy-keys spacemacs/ace-buffer-links (avy--process (spacemacs//collect-spacemacs-buffer-links) (function avy--overlay-pre))))) (if res (progn (goto-char (1+ res)) (widget-button-press (point)))))
  spacemacs/ace-buffer-links()
  call-interactively(spacemacs/ace-buffer-links nil nil)
  command-execute(spacemacs/ace-buffer-links)
```
The problem seems to be that an installed package no longer exists, hence its old version is "20150307.208" and its "new" version is nil.

Suspiciously, I also found out that the [highlight-numbers](https://github.com/Fanael/highlight-numbers) package is currently not available from MELPA, although the README says it should be.
This link http://melpa.org/#/highlight-numbers tells me that the package was not found.
However, the "20150307.208" version string can't belong to `highlight-numbers` - it doesn't match any of its commit dates (see [here](https://github.com/Fanael/highlight-numbers/commits/master)).

I might be able to do more research later.

EDIT: by the time I finished writing this comment, the `highlight-numbers` package has become available again from MELPA. Fortunately I have some (weak) proof in these screenshots:
Before:
![package-not-available](https://cloud.githubusercontent.com/assets/4334375/9486669/f0f20260-4bd2-11e5-845b-1c34010c6c84.png)
After:
![package-available](https://cloud.githubusercontent.com/assets/4334375/9486670/f60b81b8-4bd2-11e5-8fa7-1b78215ecf28.png)

**Comment by bmag** [comment]
I was able to solve the issue by adding `toxi-theme` to `dotspacemacs-excluded-packages`, then restarting emacs. The toxi-theme package is included in the themes-megapack layer, which I have enabled.

According to toxi-theme's entry in [libraries.io](https://libraries.io/emacs/toxi-theme), it should be available from MELPA (http://melpa.org/#/toxi-theme), and the source of the package should be located at http://hg.postspectacular.com/toxi-theme/
However, the package is currently not available from MELPA, and the source code is not available either.

So my conclusion is that this bug caused by an installed package being temporarily (or permanently) unavailable.
The function `configuration-layer//new-version-available-p` could detect this (`new-version`'s value will be `nil`) and emit a warning or something.
